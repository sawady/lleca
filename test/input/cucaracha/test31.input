

fun p(x : Bool) {
    if x {
        putChar(84)
        putChar(114)
        putChar(117)
        putChar(101)
    } else {
        putChar(70)
        putChar(97)
        putChar(108)
        putChar(115)
        putChar(101)
    }
}

fun sep(op : Int) {
    putChar(32)
    if op == 1 {
      putChar(97)
      putChar(110)
      putChar(100)
    } else {
      putChar(111)
      putChar(114)
    }
    putChar(32)
}

fun eq() {
    putChar(32)
    putChar(61)
    putChar(32)
}

fun enter() {
    putChar(10)
}

fun fila(x : Bool, op1 : Int, y : Bool, op2 : Int, z : Bool) {

    if op1 == 1 {
        r := x and y
    } else {
        r := x or y
    }

    if op2 == 1 {
        r := r and z
    } else {
        r := r or z
    }

    p(x) sep(op1) p(y) sep(op2) p(z) eq() p(r) enter()
}

fun tabla(op1 : Int, op2 : Int) {
  putChar(45) putChar(45) putChar(45) putChar(10)
  fila(False, op1, False, op2, False)
  fila(False, op1, False, op2, True)
  fila(False, op1, True,  op2, False)
  fila(False, op1, True,  op2, True)
  fila(True,  op1, False, op2, False)
  fila(True,  op1, False, op2, True)
  fila(True,  op1, True,  op2, False)
  fila(True,  op1, True,  op2, True)
  putChar(45) putChar(45) putChar(45) putChar(10)
}

fun main() {
  tabla(1, 1)
  tabla(1, 2)
  tabla(2, 1)
  tabla(2, 2)
}

