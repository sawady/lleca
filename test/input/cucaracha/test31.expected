Cons(
  Function(
    p,
    Cons(
      Parameter(
        x,
        Bool
      ),
      Nil
    ),
    Unit,
    Cons(
      StmtIf(
        ExprVar(
          x
        ),
        Cons(
          StmtCall(
            putChar,
            Cons(
              ExprConstNum(
                84
              ),
              Nil
            )
          ),
          Cons(
            StmtCall(
              putChar,
              Cons(
                ExprConstNum(
                  114
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    117
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      101
                    ),
                    Nil
                  )
                ),
                Nil
              )
            )
          )
        ),
        Just(
          Cons(
            StmtCall(
              putChar,
              Cons(
                ExprConstNum(
                  70
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    97
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      108
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        115
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      putChar,
                      Cons(
                        ExprConstNum(
                          101
                        ),
                        Nil
                      )
                    ),
                    Nil
                  )
                )
              )
            )
          )
        )
      ),
      Nil
    )
  ),
  Cons(
    Function(
      sep,
      Cons(
        Parameter(
          op,
          Int
        ),
        Nil
      ),
      Unit,
      Cons(
        StmtCall(
          putChar,
          Cons(
            ExprConstNum(
              32
            ),
            Nil
          )
        ),
        Cons(
          StmtIf(
            ExprEq(
              ExprVar(
                op
              ),
              ExprConstNum(
                1
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    97
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      110
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        100
                      ),
                      Nil
                    )
                  ),
                  Nil
                )
              )
            ),
            Just(
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      111
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        114
                      ),
                      Nil
                    )
                  ),
                  Nil
                )
              )
            )
          ),
          Cons(
            StmtCall(
              putChar,
              Cons(
                ExprConstNum(
                  32
                ),
                Nil
              )
            ),
            Nil
          )
        )
      )
    ),
    Cons(
      Function(
        eq,
        Nil,
        Unit,
        Cons(
          StmtCall(
            putChar,
            Cons(
              ExprConstNum(
                32
              ),
              Nil
            )
          ),
          Cons(
            StmtCall(
              putChar,
              Cons(
                ExprConstNum(
                  61
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    32
                  ),
                  Nil
                )
              ),
              Nil
            )
          )
        )
      ),
      Cons(
        Function(
          enter,
          Nil,
          Unit,
          Cons(
            StmtCall(
              putChar,
              Cons(
                ExprConstNum(
                  10
                ),
                Nil
              )
            ),
            Nil
          )
        ),
        Cons(
          Function(
            fila,
            Cons(
              Parameter(
                x,
                Bool
              ),
              Cons(
                Parameter(
                  op1,
                  Int
                ),
                Cons(
                  Parameter(
                    y,
                    Bool
                  ),
                  Cons(
                    Parameter(
                      op2,
                      Int
                    ),
                    Cons(
                      Parameter(
                        z,
                        Bool
                      ),
                      Nil
                    )
                  )
                )
              )
            ),
            Unit,
            Cons(
              StmtIf(
                ExprEq(
                  ExprVar(
                    op1
                  ),
                  ExprConstNum(
                    1
                  )
                ),
                Cons(
                  StmtAssign(
                    r,
                    ExprAnd(
                      ExprVar(
                        x
                      ),
                      ExprVar(
                        y
                      )
                    )
                  ),
                  Nil
                ),
                Just(
                  Cons(
                    StmtAssign(
                      r,
                      ExprOr(
                        ExprVar(
                          x
                        ),
                        ExprVar(
                          y
                        )
                      )
                    ),
                    Nil
                  )
                )
              ),
              Cons(
                StmtIf(
                  ExprEq(
                    ExprVar(
                      op2
                    ),
                    ExprConstNum(
                      1
                    )
                  ),
                  Cons(
                    StmtAssign(
                      r,
                      ExprAnd(
                        ExprVar(
                          r
                        ),
                        ExprVar(
                          z
                        )
                      )
                    ),
                    Nil
                  ),
                  Just(
                    Cons(
                      StmtAssign(
                        r,
                        ExprOr(
                          ExprVar(
                            r
                          ),
                          ExprVar(
                            z
                          )
                        )
                      ),
                      Nil
                    )
                  )
                ),
                Cons(
                  StmtCall(
                    p,
                    Cons(
                      ExprVar(
                        x
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      sep,
                      Cons(
                        ExprVar(
                          op1
                        ),
                        Nil
                      )
                    ),
                    Cons(
                      StmtCall(
                        p,
                        Cons(
                          ExprVar(
                            y
                          ),
                          Nil
                        )
                      ),
                      Cons(
                        StmtCall(
                          sep,
                          Cons(
                            ExprVar(
                              op2
                            ),
                            Nil
                          )
                        ),
                        Cons(
                          StmtCall(
                            p,
                            Cons(
                              ExprVar(
                                z
                              ),
                              Nil
                            )
                          ),
                          Cons(
                            StmtCall(
                              eq,
                              Nil
                            ),
                            Cons(
                              StmtCall(
                                p,
                                Cons(
                                  ExprVar(
                                    r
                                  ),
                                  Nil
                                )
                              ),
                              Cons(
                                StmtCall(
                                  enter,
                                  Nil
                                ),
                                Nil
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          ),
          Cons(
            Function(
              tabla,
              Cons(
                Parameter(
                  op1,
                  Int
                ),
                Cons(
                  Parameter(
                    op2,
                    Int
                  ),
                  Nil
                )
              ),
              Unit,
              Cons(
                StmtCall(
                  putChar,
                  Cons(
                    ExprConstNum(
                      45
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        45
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      putChar,
                      Cons(
                        ExprConstNum(
                          45
                        ),
                        Nil
                      )
                    ),
                    Cons(
                      StmtCall(
                        putChar,
                        Cons(
                          ExprConstNum(
                            10
                          ),
                          Nil
                        )
                      ),
                      Cons(
                        StmtCall(
                          fila,
                          Cons(
                            ExprConstBool(
                              False
                            ),
                            Cons(
                              ExprVar(
                                op1
                              ),
                              Cons(
                                ExprConstBool(
                                  False
                                ),
                                Cons(
                                  ExprVar(
                                    op2
                                  ),
                                  Cons(
                                    ExprConstBool(
                                      False
                                    ),
                                    Nil
                                  )
                                )
                              )
                            )
                          )
                        ),
                        Cons(
                          StmtCall(
                            fila,
                            Cons(
                              ExprConstBool(
                                False
                              ),
                              Cons(
                                ExprVar(
                                  op1
                                ),
                                Cons(
                                  ExprConstBool(
                                    False
                                  ),
                                  Cons(
                                    ExprVar(
                                      op2
                                    ),
                                    Cons(
                                      ExprConstBool(
                                        True
                                      ),
                                      Nil
                                    )
                                  )
                                )
                              )
                            )
                          ),
                          Cons(
                            StmtCall(
                              fila,
                              Cons(
                                ExprConstBool(
                                  False
                                ),
                                Cons(
                                  ExprVar(
                                    op1
                                  ),
                                  Cons(
                                    ExprConstBool(
                                      True
                                    ),
                                    Cons(
                                      ExprVar(
                                        op2
                                      ),
                                      Cons(
                                        ExprConstBool(
                                          False
                                        ),
                                        Nil
                                      )
                                    )
                                  )
                                )
                              )
                            ),
                            Cons(
                              StmtCall(
                                fila,
                                Cons(
                                  ExprConstBool(
                                    False
                                  ),
                                  Cons(
                                    ExprVar(
                                      op1
                                    ),
                                    Cons(
                                      ExprConstBool(
                                        True
                                      ),
                                      Cons(
                                        ExprVar(
                                          op2
                                        ),
                                        Cons(
                                          ExprConstBool(
                                            True
                                          ),
                                          Nil
                                        )
                                      )
                                    )
                                  )
                                )
                              ),
                              Cons(
                                StmtCall(
                                  fila,
                                  Cons(
                                    ExprConstBool(
                                      True
                                    ),
                                    Cons(
                                      ExprVar(
                                        op1
                                      ),
                                      Cons(
                                        ExprConstBool(
                                          False
                                        ),
                                        Cons(
                                          ExprVar(
                                            op2
                                          ),
                                          Cons(
                                            ExprConstBool(
                                              False
                                            ),
                                            Nil
                                          )
                                        )
                                      )
                                    )
                                  )
                                ),
                                Cons(
                                  StmtCall(
                                    fila,
                                    Cons(
                                      ExprConstBool(
                                        True
                                      ),
                                      Cons(
                                        ExprVar(
                                          op1
                                        ),
                                        Cons(
                                          ExprConstBool(
                                            False
                                          ),
                                          Cons(
                                            ExprVar(
                                              op2
                                            ),
                                            Cons(
                                              ExprConstBool(
                                                True
                                              ),
                                              Nil
                                            )
                                          )
                                        )
                                      )
                                    )
                                  ),
                                  Cons(
                                    StmtCall(
                                      fila,
                                      Cons(
                                        ExprConstBool(
                                          True
                                        ),
                                        Cons(
                                          ExprVar(
                                            op1
                                          ),
                                          Cons(
                                            ExprConstBool(
                                              True
                                            ),
                                            Cons(
                                              ExprVar(
                                                op2
                                              ),
                                              Cons(
                                                ExprConstBool(
                                                  False
                                                ),
                                                Nil
                                              )
                                            )
                                          )
                                        )
                                      )
                                    ),
                                    Cons(
                                      StmtCall(
                                        fila,
                                        Cons(
                                          ExprConstBool(
                                            True
                                          ),
                                          Cons(
                                            ExprVar(
                                              op1
                                            ),
                                            Cons(
                                              ExprConstBool(
                                                True
                                              ),
                                              Cons(
                                                ExprVar(
                                                  op2
                                                ),
                                                Cons(
                                                  ExprConstBool(
                                                    True
                                                  ),
                                                  Nil
                                                )
                                              )
                                            )
                                          )
                                        )
                                      ),
                                      Cons(
                                        StmtCall(
                                          putChar,
                                          Cons(
                                            ExprConstNum(
                                              45
                                            ),
                                            Nil
                                          )
                                        ),
                                        Cons(
                                          StmtCall(
                                            putChar,
                                            Cons(
                                              ExprConstNum(
                                                45
                                              ),
                                              Nil
                                            )
                                          ),
                                          Cons(
                                            StmtCall(
                                              putChar,
                                              Cons(
                                                ExprConstNum(
                                                  45
                                                ),
                                                Nil
                                              )
                                            ),
                                            Cons(
                                              StmtCall(
                                                putChar,
                                                Cons(
                                                  ExprConstNum(
                                                    10
                                                  ),
                                                  Nil
                                                )
                                              ),
                                              Nil
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            ),
            Cons(
              Function(
                main,
                Nil,
                Unit,
                Cons(
                  StmtCall(
                    tabla,
                    Cons(
                      ExprConstNum(
                        1
                      ),
                      Cons(
                        ExprConstNum(
                          1
                        ),
                        Nil
                      )
                    )
                  ),
                  Cons(
                    StmtCall(
                      tabla,
                      Cons(
                        ExprConstNum(
                          1
                        ),
                        Cons(
                          ExprConstNum(
                            2
                          ),
                          Nil
                        )
                      )
                    ),
                    Cons(
                      StmtCall(
                        tabla,
                        Cons(
                          ExprConstNum(
                            2
                          ),
                          Cons(
                            ExprConstNum(
                              1
                            ),
                            Nil
                          )
                        )
                      ),
                      Cons(
                        StmtCall(
                          tabla,
                          Cons(
                            ExprConstNum(
                              2
                            ),
                            Cons(
                              ExprConstNum(
                                2
                              ),
                              Nil
                            )
                          )
                        ),
                        Nil
                      )
                    )
                  )
                )
              ),
              Nil
            )
          )
        )
      )
    )
  )
)
